### PR TITLE
update to cython 3

### PR DESCRIPTION
**Pull request by emlys**
Update to Cython 3. Also fixes the github actions workflow for the new mac runners.

### CODE REVIEW

File: .github/workflows/build-py-dists.yml [hunk 1]
```diff
@@ -8,8 +8,7 @@ jobs:
             fail-fast: false
             matrix:
                 os: [windows-latest, macos-latest]
-                python-version: [3.7, 3.8, 3.9, "3.10", "3.11", "3.12"]
-                python-arch: [x64]
```
Comment (emlys) [inline .github/workflows/build-py-dists.yml:12]: `python-arch` isn't needed here, and it caused an issue with the new ARM64 mac runners.

File: .github/workflows/build-py-dists.yml [hunk 2]
```diff
@@ -8,8 +8,7 @@ jobs:
             fail-fast: false
             matrix:
                 os: [windows-latest, macos-latest]
-                python-version: [3.7, 3.8, 3.9, "3.10", "3.11", "3.12"]
```
Comment (emlys) [inline .github/workflows/build-py-dists.yml:11]: python 3.7 is end-of-life, we already removed it from the test workflow. And `setup-python` couldn't find an ARM64 python 3.7 distribution.

File: .github/workflows/build-py-dists.yml (hunk 1 above)
Comment (phargogh) [reply .github/workflows/build-py-dists.yml:12]: I bet this is a holdover from when we were still doing x86 builds.  Makes sense to remove!

File: HISTORY.rst [hunk 3]
```diff
@@ -10,6 +10,7 @@ Unreleased Changes
   called with an invalid resampling algorithm. We now fall back to the
   underlying GDAL functions' error messages.
   https://github.com/natcap/pygeoprocessing/issues/387
+* Updated to Cython 3.
```
Comment (phargogh) [inline HISTORY.rst:13]: If we are removing testing/wheel-building support for python 3.7, should that also be noted here in HISTORY and also removed from the `setup.py` trove classifiers?

**Review by phargogh — CHANGES_REQUESTED** [review]
Looks great to me, I just had one comment about dropping explicit support for python 3.7.

File: HISTORY.rst (hunk 3 above)
Comment (emlys) [reply HISTORY.rst:13]: Ooh you're right

**Review by phargogh — APPROVED** [review]
Code changes look good to me, thanks @emlys !  Just waiting for the tests to pass.

**Pull request merged**